File: 3Sum/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()  # O(nlogn)
        for i in range(len(nums)):  # O(n)
            if i > 0 and nums[i] == nums[i-1]:
                continue
            left = i+1
            right = len(nums)-1
            while left < right:  # O(n)
                s = nums[i] + nums[left] + nums[right]
                if s > 0:
                    right -= 1
                elif s < 0:
                    left += 1
                else:
                    result.append([nums[i], nums[left], nums[right]])
                    left += 1
                    while left < right and nums[left] == nums[left-1]:
                        left += 1
        return result

# time complexity: O(n^2)

File: 3Sum/3Sum.cpp
class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        vector<vector<int>> result;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size(); i++)
        {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            int left = i + 1;
            int right = nums.size() - 1;
            while (left < right)
            {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum > 0)
                {
                    right -= 1;
                }
                else if (sum < 0)
                {
                    left += 1;
                }
                else
                {
                    result.push_back({nums[i], nums[left], nums[right]});
                    left += 1;
                    while (left < right && nums[left] == nums[left - 1])
                    {
                        left += 1;
                    }
                }
            }
        }
        return result;
    }
};